From da85eae04fac4e3bd3eecf9c558866bae9e2c3f1 Mon Sep 17 00:00:00 2001
From: Dave Reisner <d@falconindy.com>
Date: Tue, 26 Apr 2011 12:03:08 -0400
Subject: [PATCH] avoid calling yajl_complete_parse when curl fails

I don't entirely like the fix applied for yajl GH #27, because it
requires that cower's json parser set yajl_allow_partial_values to end
cleanly. Since this is conflicting, I'm fixing this on the application
side as well.
---
 cower.c |    6 +++---
 1 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/cower.c b/cower.c
index 056a862..a4f6b26 100644
--- a/cower.c
+++ b/cower.c
@@ -1889,9 +1889,6 @@ void *task_query(CURL *curl, void *arg) { /* {{{ */
   cwr_printf(LOG_DEBUG, "[%p]: curl_easy_perform %s\n", (void*)pthread_self(), url);
   curlstat = curl_easy_perform(curl);
 
-  yajl_complete_parse(yajl_hand);
-  yajl_free(yajl_hand);
-
   if (curlstat != CURLE_OK) {
     cwr_fprintf(stderr, LOG_ERROR, "curl: %s\n", curl_easy_strerror(curlstat));
     goto finish;
@@ -1903,6 +1900,8 @@ void *task_query(CURL *curl, void *arg) { /* {{{ */
     goto finish;
   }
 
+  yajl_complete_parse(yajl_hand);
+
   pkglist = parse_struct->pkglist;
 
   if (pkglist && cfg.extinfo) {
@@ -1925,6 +1924,7 @@ void *task_query(CURL *curl, void *arg) { /* {{{ */
   }
 
 finish:
+  yajl_free(yajl_hand);
   curl_free(escaped);
   FREE(parse_struct);
   FREE(url);
-- 
1.7.4.4

